--- before/net/minecraft/client/gui/GuiCommandBlock.java
+++ after/net/minecraft/client/gui/GuiCommandBlock.java
@@ -19,6 +19,7 @@
 public class GuiCommandBlock extends GuiScreen implements ITabCompleter
 {
     private GuiTextField commandTextField;
+    private com.cleanroommc.client.chat.suggestion.CommandSuggestions suggestions;
     private GuiTextField previousOutputTextField;
     private final TileEntityCommandBlock commandBlock;
     private GuiButton doneBtn;
@@ -85,6 +86,7 @@
                 return commandblockbaselogic.getPosition();
             }
         };
+        this.suggestions = new com.cleanroommc.client.chat.suggestion.CommandSuggestions(this.commandTextField, this.tabCompleter, true);
     }
 
     public void updateGui()
@@ -170,6 +172,7 @@
     {
         this.tabCompleter.resetRequested();
 
+        if (this.suggestions.keyTyped(keyCode)) return;
         if (keyCode == 15)
         {
             this.tabCompleter.complete();
@@ -181,6 +184,7 @@
 
         this.commandTextField.textboxKeyTyped(typedChar, keyCode);
         this.previousOutputTextField.textboxKeyTyped(typedChar, keyCode);
+        this.suggestions.refresh();
 
         if (keyCode == 28 || keyCode == 156)
         {
@@ -195,11 +199,19 @@
     @Override
     protected void mouseClicked(int mouseX, int mouseY, int mouseButton) throws IOException
     {
+        if (this.suggestions.mouseClicked(mouseX, mouseY, mouseButton)) return;
         super.mouseClicked(mouseX, mouseY, mouseButton);
         this.commandTextField.mouseClicked(mouseX, mouseY, mouseButton);
         this.previousOutputTextField.mouseClicked(mouseX, mouseY, mouseButton);
+        this.suggestions.refresh();
     }
 
+    @Override
+    public void handleMouseInput() throws IOException {
+        super.handleMouseInput();
+        this.suggestions.handleMouseInput(this);
+    }
+
     @Override
     public void drawScreen(int mouseX, int mouseY, float partialTicks)
     {
@@ -206,7 +218,7 @@
         this.drawDefaultBackground();
         this.drawCenteredString(this.fontRenderer, I18n.format("advMode.setCommand"), this.width / 2, 20, 16777215);
         this.drawString(this.fontRenderer, I18n.format("advMode.command"), this.width / 2 - 150, 40, 10526880);
-        this.commandTextField.drawTextBox();
+        this.suggestions.drawTextBox();
         int i = 75;
         int j = 0;
         this.drawString(
@@ -232,6 +244,7 @@
             this.previousOutputTextField.drawTextBox();
         }
 
+        this.suggestions.drawDropdown(mouseX, mouseY);
         super.drawScreen(mouseX, mouseY, partialTicks);
     }
 
@@ -313,5 +326,6 @@
     public void setCompletions(String... newCompletions)
     {
         this.tabCompleter.setCompletions(newCompletions);
+        this.suggestions.setCompletions(newCompletions);
     }
 }
